lib/charms/legend_database_manager/v0/database_manager.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2021 Canonical
+# See LICENSE file for licensing details.
+
+import json
+
+from ops import framework
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "431732f8afb641a3a5a38e5c5d01ee11"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 0
+
+REQUIRED_LEGEND_DATABASE_CREDENTIALS = [
+    "username", "password", "database", "uri"]
+
+
+class LegendDatabaseConsumer(framework.Object):
+    def __init__(self, charm, relation_name="legend-db"):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+
+    def database_connection_from_mongodb(
+            self, mongodb_consumer, mongodb_relation_id):
+        """Returns a dict with Mongo connection info for Legend components
+        just like `get_legend_database_creds`.
+
+        Args:
+            mongodb_consumer:  charms.mongodb_k8s.v0.mongodb.MongoConsumer
+            mongodb_relation_id: ID of the MongoDB relation.
+
+        Returns:
+            Dictionary with the following structure:
+            {
+                "uri": "<replica set URI (with user/pass, no DB name)>",
+                "username": "<username>",
+                "password": "<password>",
+                "database": "<database name>"
+            }
+        """
+        creds = mongodb_consumer.credentials(rel_id=mongodb_relation_id)
+        if not creds:
+            return {}
+
+        databases = mongodb_consumer.databases(rel_id=mongodb_relation_id)
+        if not databases:
+            return {}
+
+        uri = creds['replica_set_uri']
+        # NOTE: we remove the trailing database from the URI:
```

Comment:
Double check me on this but I think the database in the replica set URI specifies the authentication database. If so you may leave it as is.

Reply:
I'm not entirely sure on which data database is specified within the URI, but from my testing, the Legend components (or whatever Java Mongo lib they're all using) don't like the database name appended to the URI.

As mentioned, this method is here to "adapt" the info returned by the MongoConsumer into the format the relation between the LDBM and other Legend components uses.